### PR TITLE
Split jnis into 2 libs and add common lib

### LINKED ISSUE #173: Splitting up nmslib jni library from faiss
Currently, we compile  nmslib and faiss into the same JNI library. While this works, it couples the libraries together unnecessarily. If a user is just interested in nmslib, they will need to load the faiss portion of the library into memory as well. faiss adds some additional dependencies (OpenMP) which may cause compatibility issues. 

Separating the 2 libraries into 2 jni's will remove this coupling. Because we dynamically load the JNI library into memory only when it is needed, a user only using nmslib will never have to worry about loading the faiss portion of the library into memory.

Additionally, to reduce code redundancy, we can compile a third commons library that can be used to share common functionality.

### PR DESCRIPTION
**Pull request by jmazanec15**
Signed-off-by: John Mazanec <jmazane@amazon.com>

### Description
This PR splits the "opensearchknn" jni into 2 jni's: "opensearchknn_nmslib" and "opensearchknn_faiss". In addition, it builds another shared library "opensearchknn_common" that will be shared between the 2. The primary reason for splitting the 2 libraries is isolation: for a user using nmslib only, they should not need to load faiss and vice-versa.

Previously, routing to the correct engine was done in `jni/src/org_opensearch_knn_index_JNIService.cpp`. This change moves this routing decision up into the Java layer. JNIService will route request to either FaissService or NmslibService.

This PR is in WIP for the following:
- [x] Ensure Docker image and tar ball do not break
- [x] Fix CPack logic
- [x] A few minor tweaks

I will remove WIP status once I think it is ready to merge.
 
### Issues Resolved
#173 
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jmazanec15** [comment]
Confirmed tar works arm/x64 on AL2 and  Ubuntu 18. Confirmed docker works on arm and x64. PR is ready to be reviewed/shipped.

File: jni/CMakeLists.txt [hunk 1]
```diff
@@ -28,12 +28,25 @@ cmake_minimum_required(VERSION 3.17)
...
 
+option(CONFIG_FAISS "Configure faiss library build when this is on")
+option(CONFIG_NMSLIB "Configure nmslib library build when this is on")
+option(CONFIG_TEST "Configure tests when this is on")
+
+if (${CONFIG_FAISS} STREQUAL OFF AND ${CONFIG_NMSLIB} STREQUAL OFF AND ${CONFIG_TEST} STREQUAL OFF)
+    set(CONFIG_ALL ON)
+else()
+    set(CONFIG_ALL OFF)
+endif ()
```
Comment (vamshin) [inline jni/CMakeLists.txt:48]:
```suggestion
if (${CONFIG_FAISS} STREQUAL ON OR ${CONFIG_NMSLIB} STREQUAL ON OR ${CONFIG_TEST} STREQUAL ON)
    set(CONFIG_ALL OFF)
endif ()
```

Comment (jmazanec15) [reply jni/CMakeLists.txt:48]: But then if they are all on, config_all will be set to off.

Comment (vamshin) [reply jni/CMakeLists.txt:48]: Good catch!

**Review by vamshin — APPROVED** [review]
LGTM! Thanks

**Pull request merged**